In [0]:
import os
#read_path = 'D:/NYU-1/DL/data/chest_xray'
test_normal = 'D:/NYU-1/DL/project/chest_xray/test/NORMAL'
test_pneumonia = 'D:/NYU-1/DL/project/chest_xray/test/PNEUMONIA'



# 2 classes: normal, pneumonia

### test_normal to test_normal and val_normal

In [0]:
os.chdir(test_normal) #Change path 
normal_name_list2 = os.listdir() #Get file name 
#csv_missing=[]

In [0]:
test_n = []
for i in normal_name_list2:
    if 'NORMAL2' in i:
        i = i[8:]
    patient = i[:7]
    test_n.append(patient)
test_normal_distinct = list(set(test_n))
print('numer of images in test_normal:',len(test_n))
print('number of patients in test_normal:',len(test_normal_distinct))   #patient number

numer of images in test_normal: 234
number of patients in test_normal: 202


In [0]:
print(test_normal_distinct[:5])
import random
random.shuffle(test_normal_distinct)
print(test_normal_distinct[:5])

['IM-0328', 'IM-0353', 'IM-0015', 'IM-0112', 'IM-0313']
['IM-0091', 'IM-0081', 'IM-0075', 'IM-0325', 'IM-0123']


In [0]:
val_n_id = test_normal_distinct[:101]
test_n_id = test_normal_distinct[101:]

In [0]:
test_n_list = [] #image name

for i in test_n_id:
    for j in normal_name_list2:
        if i in j:
            test_n_list.append(j)
    
print('number of images in test(split)',len(test_n_list))

number of images in test(split) 114


In [0]:
val_n_list = [] #image name

for i in val_n_id:
    for j in normal_name_list2:
        if i in j:
            val_n_list.append(j)
            
print('number of images in val(split)', len(val_n_list))

number of images in val(split) 120


### test_pneumonia to test_pneumonia and val_pneumonia

In [0]:
os.chdir(test_pneumonia) #Change path 
normal_name_list4 = os.listdir() #Get file name 
#csv_missing=[]

In [0]:
test_p = []
for i in normal_name_list4:
    patient = i.split('_')[0]
    test_p.append(patient)
    #train_normal.append(patient)
test_pneumonia_distinct = list(set(test_p))
print('number of images in test_penumonia:',len(test_p))
print('number of patients in test_pneumonia:',len(test_pneumonia_distinct))

number of images in test_penumonia: 390
number of patients in test_pneumonia: 202


In [0]:
print(test_pneumonia_distinct[:5])
random.shuffle(test_pneumonia_distinct)
print(test_pneumonia_distinct[:5])

['person41', 'person1616', 'person101', 'person173', 'person94']
['person11', 'person1670', 'person1656', 'person41', 'person71']


In [0]:
for i in range(len(test_pneumonia_distinct)):
    id_ = test_pneumonia_distinct[i]+'_'
    test_pneumonia_distinct[i] = id_
print(test_pneumonia_distinct[5])
    
    

person1637_


In [0]:
test_p_id = test_pneumonia_distinct[:101] #50%
val_p_id = test_pneumonia_distinct[101:] 

In [0]:
test_p_list = [] #image name

for i in test_p_id:
    for j in normal_name_list4:
        if i in j:
            test_p_list.append(j)
            
print('number of images in test(split)',len(test_p_list))

number of images in test(split) 196


In [0]:
val_p_list = [] #image name

for i in val_p_id:
    for j in normal_name_list4:
        if i in j:
            val_p_list.append(j)
            
print('number of images in val(split)',len(val_p_list))

number of images in val(split) 194


### train

In [0]:
train_normal = 'D:/NYU-1/DL/project/chest_xray/train/NORMAL'
train_pneumonia = 'D:/NYU-1/DL/project/chest_xray/train/PNEUMONIA'

In [0]:
os.chdir(train_normal) #Change path 
train_n_list = os.listdir() #Get file name 
#csv_missing=[]
#print(len(train_n_split))

In [0]:
train_n = []
for i in train_n_list:
    if 'NORMAL2' in i:
        i = i[8:]
    patient = i[:7]
    train_n.append(patient)
train_normal_distinct = list(set(train_n))
print('numer of images in train_normal:',len(train_n))
print('number of patients in train_normal:',len(train_normal_distinct))   #patient number

numer of images in train_normal: 1349
number of patients in train_normal: 1008


In [0]:
os.chdir(train_pneumonia) #Change path 
train_p_list = os.listdir() #Get file name 
#csv_missing=[]
#print(len(train_p_split))

In [0]:
train_p = []
for i in train_p_list:
    patient = i.split('_')[0]
    train_p.append(patient)
train_pneumonia_distinct = list(set(train_p))
print('number of images in train_penumonia:',len(train_p))
print('number of patients in train_pneumonia:',len(train_pneumonia_distinct))

number of images in train_penumonia: 3886
number of patients in train_pneumonia: 1645


### new table: indicating train/val/test 

In [0]:
import pandas as pd
data1 = {'image_index':train_p_list,
         'class':'pneumonia',
         'label': 1}
data2 = {'image_index':train_n_list,
         'class':'normal',
         'label': 0}
df1 = pd.DataFrame(data1)
df2 = pd.DataFrame(data2)
df_train = pd.concat([df1, df2], axis = 0)
print(df_train.shape)
df_train = df_train.sample(frac = 1)
df_train.head()


(5232, 3)


image_index      class  label
329          IM-0517-0001-0001.jpeg     normal      0
1399  person1666_bacteria_4416.jpeg  pneumonia      1
1860   person292_bacteria_1378.jpeg  pneumonia      1
3752   person933_bacteria_2858.jpeg  pneumonia      1
3481   person764_bacteria_2668.jpeg  pneumonia      1

In [0]:
import pandas as pd
data3 = {'image_index':test_p_list,
         'class':'pneumonia',
         'label': 1}
data4 = {'image_index':test_n_list,
         'class':'normal',
         'label': 0}
df3 = pd.DataFrame(data3)
df4 = pd.DataFrame(data4)
df_test = pd.concat([df3, df4], axis = 0)
df_test = df_test.sample(frac = 1)
df_test.shape

(310, 3)

In [0]:
data5 = {'image_index':val_p_list,
         'class':'pneumonia',
         'label': 1}
data6 = {'image_index':val_n_list,
         'class':'normal',
         'label': 0}
df5 = pd.DataFrame(data5)
df6 = pd.DataFrame(data6)
df_val = pd.concat([df5, df6], axis = 0)
df_val = df_val.sample(frac = 1)
df_val.shape

(314, 3)

In [0]:
df_train.to_csv('train.csv', index = False)
df_val.to_csv('val.csv', index = False)
df_test.to_csv('test.csv', index = False)

**train:**
normal: 1349 images, 1008 patients

pneumonia: 3886 images, 1645 patients


**test:**

normal: 114 images, 101 patients

pneumonia: 196 images, 101 patients


**validation:**

normal: 120 images, 101 patients

pneumonia: 194 images, 101 patients

**test/validation splitting based on patient id**

# 3 classes: normal, virus, bacteria

In [0]:
import pandas as pd
def class_3(file_name):
    df = pd.read_csv(file_name)

    class_3 = []
    label_3 = []
    for i in range(len(df)):
        if 'virus' in df['image_index'][i]:
            class_ = 'virus'
            label_ = 1
        elif 'bacteria' in df['image_index'][i]:
            class_ = 'bacteria'
            label_ = 2
        else:
            class_ = 'normal'
            label_ = 0
        class_3.append(class_)
        label_3.append(label_)

    image_index = list(df['image_index'])
    df_temp = {'image_index': image_index,
                'class': class_3,
                'label': label_3}
    df_new  = pd.DataFrame(df_temp)

    return df_new

In [59]:
train_new = class_3('train.csv')
print(train_new.shape)
print(train_new.head())

val_new = class_3('val.csv')
print(val_new.shape)
print(val_new.head())

test_new = class_3('test.csv')
print(test_new.shape)
print(test_new.head())

(5232, 3)
                     image_index     class  label
0         IM-0517-0001-0001.jpeg    normal      0
1  person1666_bacteria_4416.jpeg  bacteria      2
2   person292_bacteria_1378.jpeg  bacteria      2
3   person933_bacteria_2858.jpeg  bacteria      2
4   person764_bacteria_2668.jpeg  bacteria      2
(314, 3)
                   image_index     class  label
0    NORMAL2-IM-0059-0001.jpeg    normal      0
1  person134_bacteria_641.jpeg  bacteria      2
2  person134_bacteria_643.jpeg  bacteria      2
3   person99_bacteria_473.jpeg  bacteria      2
4   person1625_virus_2817.jpeg     virus      1
(310, 3)
                      image_index     class  label
0      person1651_virus_2855.jpeg     virus      1
1     person100_bacteria_478.jpeg  bacteria      2
2  NORMAL2-IM-0246-0001-0001.jpeg    normal      0
3       NORMAL2-IM-0359-0001.jpeg    normal      0
4       NORMAL2-IM-0219-0001.jpeg    normal      0


In [0]:
train_new.to_csv('train_3.csv')
val_new.to_csv('val_3.csv')
test_new.to_csv('test_3.csv')

In [64]:
print('train:\n',train_new['class'].value_counts())
print('validation:\n',val_new['class'].value_counts())
print('test:\n',test_new['class'].value_counts())

train:
 bacteria    2538
normal      1349
virus       1345
Name: class, dtype: int64
validation:
 normal      120
bacteria    114
virus        80
Name: class, dtype: int64
test:
 bacteria    128
normal      114
virus        68
Name: class, dtype: int64
